lib/THC/THCReduceApplyUtils.cuh

```diff
@@ -72,6 +72,32 @@ __device__ T reduceBlock(T* smem,
   return r;
 }
 
+// Block-wide reduction where each thread locally reduces N
+// values before letting a single warp take over
+template <typename T, typename ReduceOp, int N, bool SyncAfterLocalReduction>
+__device__ T reduceBlockN(T *smem,
+                         T threadVals[N],
+                         int numVals,
+                         ReduceOp reduceOp,
+                         T init) {
+  int offset = threadIdx.x * N;
+  T local = offset < numVals ? threadVals[0] : init;
```

Comment:
An alternative could be assuming that numVals is exactly N * blockDim.x, so:

```
T local = threadVals[0];
#pragma unroll
for (int i = 1; i < N; ++i) {
  local = reduceOp(local, threadVals[i]);
}
```

assuming that out of bounds threads initialize all threadVals[i] to the identity element for the reduction